GIFT.pegjs

```diff
@@ -65,15 +65,27 @@ Description "Description"
 Question
   = __
     title:QuestionTitle? _
-    stem1:QuestionStem _ 
+    stem1:QuestionStem? _ 
     '{' _
     answers:(MatchingAnswers / TrueFalseAnswer / MCAnswers / NumericalAnswerType / SingleCorrectShortAnswer / EssayAnswer ) _
     '}' _
     stem2:(Comment / QuestionStem)?
     QuestionSeparator
   {
     var embedded = (stem2 != null);
-    var stem = {format:stem1.format, text:stem1.text + ( embedded ? " _____ " + stem2.text : "")};
+    
+    //avoids throwing errors from empty keys
+    //Source: https://css-tricks.com/%E2%80%8B%E2%80%8Bavoiding-those-dang-cannot-read-property-of-undefined-errors/
+    var stem1Format = stem1 && stem1.format;
+    var stem1Text = stem1 && stem1.text;
+    var stem2Format = stem2 && stem2.format;
+    var stem2Text = stem2 && stem2.text;
+    
+   	var format = stem1Format || stem2Format || "moodle";
+    var text = stem1Text || "";
+    var space = stem1Text ? " " : "";
+   	
+    var stem = {format:format, text:text + ( embedded ? space + "_____ " + stem2.text : "")};
```

Comment:
The logic of `embedded` is more clear if it were in the definition of `space` (at the `var` assignment)? I find it complicated to have to look at two places to understand.

Rename `space` to `conditionalSpace`, and include the embedded variable in its assignment.

Reply:
I'm not sure how to combine the ```embedded``` logic with ```conditionalSpace```. I see two separate conditionals here: 
1. Is there a 2nd stem? (currently ```embedded```)
2. Is there a 1st stem? (currently ```conditionalSpace```)

Combined it looks like: 
```suggestion
    var space = stem1Text && (stem2 != null) ? " " : "";
```
However, I think the intended logic is that the answer underscore "_________", exists only when stem2 exists, while the ```conditionalSpace``` only exists when stem1 exists. 

If you can, please clarify your idea on how to merge the two variables together.

I'm also happy to try other options to make this piece of code more readable. If you have any suggestions, I'm happy to hear them.

Reply:
`embedded` was intended to mean underscore (answer is "embedded" in the stem), and it is true when there's a second stem parsed. 

You only need the space if there's a first stem.

So, how about getting rid of the intermediate stuff:

```javascript
    var stem1Format = stem1 && stem1.format;
    var stem1Text = stem1 ? (stem1.text + (embedded ? " " : "")) : "";
    var stem2Format = stem2 && stem2.format;
    var stem2Text = stem2 && stem2.text;
    var format = stem1Format || stem2Format || "moodle";

    var text = stem1Text + ( embedded ? "_____ " + stem2.text : "");

    var stem = {format:format, text:text};
```

This parses a quick test on https://pegjs.org/online but I didn't run all the tests. Maybe it breaks something?

Reply:
It looks okay to me. I also gave it a test with the online generator and it seemed to work fine with and without a second stem.

You also gave me the idea to try different ways of refactoring this piece of code. I realised I added in a few unnecessary variables and we could easily write this logic in about 5 lines:

```javascript
    var embedded = (stem2 !== null);    
    var stem1Text = stem1 ? (stem1.text + (embedded ? " " : "")) : "";

    var format = (stem1 && stem1.format) || (stem2 && stem2.format) || "moodle";
    var text = stem1Text + ( embedded ? "_____ " + stem2.text : "");
    
    var question = {type:answers.type, title:title, stem: {format, text}, hasEmbeddedAnswers:embedded};
```
Since `stem1.text` and `stem2.text` are in conditionals, we don't need to wrap them in further conditionals to check if they are present. I also put the `stem1Format` and `stem2Format` directly in the format variable since we're not using them anywhere else.

We can also directly insert the `stem` variable directly in the `question` variable. In Javascript's ES6, we can use the object property shorthand, and write `{ format, text }`, rather than `{ format: format, text: text}`. However, to maintain compatibility for pre-ES6 browsers, we may need to include the keys as well.

Let me know your thoughts and I'll refactor accordingly.